### PR TITLE
MongeMapTrainer checkpoint saving and loading

### PR DESCRIPTION
**Pull request by DriessenA**
Added checkpoint saving and loading to the Monge Map trainer. I used this for the CAR exp but I think it is in general quite usefull

### CODE REVIEW

File: cmonge/trainers/ot_trainer.py [hunk 1]
```diff
@@ -179,11 +179,50 @@ def transport(self, source: jnp.ndarray) -> jnp.ndarray:
         """Transports a batch of data using the learned model."""
         return self.state.apply_fn({"params": self.state.params}, source)
 
-    def save_checkpoint(self, path: Path) -> None:
-        raise NotImplementedError
+    def save_checkpoint(self, path: Path = None, config: DotMap = None) -> None:
+        if path is None and config is None:
+            logger.error(
+                """Please provide a checkpoint save path
```
Comment (jannisborn) [inline cmonge/trainers/ot_trainer.py:187]: mention in error that model checkpoint was not saved

File: cmonge/trainers/ot_trainer.py [hunk 2]
```diff
@@ -179,11 +179,50 @@ def transport(self, source: jnp.ndarray) -> jnp.ndarray:
...
+
+    @classmethod
+    def load_checkpoint(
+        cls,
+        jobid: int,
+        logger_path: Path,
+        config: DotMap,
+        ckpt_path: Path = None,
+    ) -> None:
+        out_class = cls(
```
Comment (jannisborn) [inline cmonge/trainers/ot_trainer.py]: wrap whole method in try/except block and add a meaningful error msg

File: cmonge/trainers/ot_trainer.py [hunk 3]
```diff
@@ -179,11 +179,50 @@ def transport(self, source: jnp.ndarray) -> jnp.ndarray:
         """Transports a batch of data using the learned model."""
         return self.state.apply_fn({"params": self.state.params}, source)
 
-    def save_checkpoint(self, path: Path) -> None:
-        raise NotImplementedError
+    def save_checkpoint(self, path: Path = None, config: DotMap = None) -> None:
+        if path is None and config is None:
```
Comment (jannisborn) [inline cmonge/trainers/ot_trainer.py:185]: return here otherwise downstream code will fail

**Review by jannisborn — CHANGES_REQUESTED** [review]
That's definitely a good idea. Some changes required to robustify code, we could do the same for the conditional trainers

**Review by DriessenA — COMMENTED** [review]
Okay, I think all changes have been implemented

File: cmonge/trainers/conditional_monge_trainer.py [hunk 4]
```diff
@@ -349,15 +357,30 @@ def load_checkpoint(
...
+                config=config,
+                datamodule=datamodule,
+            )
+            if ckpt_path is None:
+                if len(config.checkpointing_path) > 0:
+                    ckpt_path = config.checkpointing_path
+                else:
+                    logger.error(
+                        """Provide checkpointing path either directly or
+                        through the model config"""
```
Comment (jannisborn) [inline cmonge/trainers/conditional_monge_trainer.py]: single quotes and one line

File: cmonge/trainers/conditional_monge_trainer.py [hunk 5]
```diff
@@ -349,15 +357,30 @@ def load_checkpoint(
...
+                    )
+                    return
+            checkpointer = PyTreeCheckpointer()
+            out_class.state_neural_net = checkpointer.restore(
+                ckpt_path, item=out_class.state_neural_net
+            )
+            logger.info("Loaded ConditionalMongeTrainer from checkpoint")
+            return out_class
+        except:
```
Comment (jannisborn) [inline cmonge/trainers/conditional_monge_trainer.py]: dont use bare except

File: cmonge/trainers/ot_trainer.py [hunk 6]
```diff
@@ -179,11 +179,57 @@ def transport(self, source: jnp.ndarray) -> jnp.ndarray:
...
+        checkpointer.save(path, ckpt, save_args=save_args, force=True)
+
+    @classmethod
+    def load_checkpoint(
+        cls,
+        jobid: int,
+        logger_path: Path,
+        config: DotMap,
+        ckpt_path: Path = None,
+    ) -> None:
```
Comment (jannisborn) [inline cmonge/trainers/ot_trainer.py]: adapt type annotation

File: cmonge/trainers/ot_trainer.py [hunk 7]
```diff
@@ -179,11 +179,57 @@ def transport(self, source: jnp.ndarray) -> jnp.ndarray:
...
+                ckpt_path, item=out_class.solver.state_neural_net
+            )
+            out_class.state = out_class.solver.state_neural_net
+            logger.info("Loaded MongeMapTrainer from checkpoint")
+            return out_class
+        except:
+            logger.error(
+                "Failed to load checkpoint, are you sure checkpoint was saved and correct path is provided?"
+            )
+            return
```
Comment (jannisborn) [inline cmonge/trainers/ot_trainer.py]: this return is useless

**Review by jannisborn — CHANGES_REQUESTED** [review]

File: cmonge/trainers/conditional_monge_trainer.py (hunk 5 above)
Comment (DriessenA) [reply cmonge/trainers/conditional_monge_trainer.py]: replaced with `except Exception`

File: cmonge/trainers/ot_trainer.py (hunk 6 above)
Comment (DriessenA) [reply cmonge/trainers/ot_trainer.py]: Return type now `Union[AbstractTrainer, None]`

File: cmonge/trainers/conditional_monge_trainer.py [hunk 8]
```diff
@@ -334,7 +334,15 @@ def evaluate_split(
 
         create_or_update_logfile(self.logger_path, self.metrics)
 
-    def save_checkpoint(self, path: Path) -> None:
+    def save_checkpoint(self, path: Path, config: DotMap = None) -> None:
+        if path is None and config is None:
+            logger.error(
+                """Please provide a checkpoint save path
+            either directly or through the config, checkpoint was NOT saved."""
+            )
```
Comment (jannisborn) [inline cmonge/trainers/conditional_monge_trainer.py]:
```suggestion
            logger.error("Please provide a checkpoint save path either directly or through the config, checkpoint was NOT saved.")
```

File: cmonge/trainers/conditional_monge_trainer.py [hunk 9]
```diff
@@ -348,16 +353,29 @@ def load_checkpoint(
...
+            out_class.state_neural_net = checkpointer.restore(
+                ckpt_path, item=out_class.state_neural_net
+            )
+            logger.info("Loaded ConditionalMongeTrainer from checkpoint")
+            return out_class
+        except Exception:
+            logger.error(
+                "Failed to load checkpoint, are you sure checkpoint was saved and correct path is provided?"
+            )
+            return
```
Comment (jannisborn) [inline cmonge/trainers/conditional_monge_trainer.py]:
```suggestion
```

File: cmonge/trainers/ot_trainer.py [hunk 10]
```diff
@@ -179,11 +181,55 @@ def transport(self, source: jnp.ndarray) -> jnp.ndarray:
...
+                ckpt_path, item=out_class.solver.state_neural_net
+            )
+            out_class.state = out_class.solver.state_neural_net
+            logger.info("Loaded MongeMapTrainer from checkpoint")
+            return out_class
+        except Exception:
+            logger.error(
+                "Failed to load checkpoint, are you sure checkpoint was saved and correct path is provided?"
+            )
+            return
```
Comment (jannisborn) [inline cmonge/trainers/ot_trainer.py]:
```suggestion
```

File: cmonge/trainers/ot_trainer.py (hunk 3 above)
Comment (jannisborn) [reply cmonge/trainers/ot_trainer.py:185]: you are still not returning

File: cmonge/trainers/ot_trainer.py (hunk 1 above)
Comment (jannisborn) [reply cmonge/trainers/ot_trainer.py:187]: again single quotes/line are enough

**Review by jannisborn — CHANGES_REQUESTED** [review]

**Comment by DriessenA** [comment]
So this code is now good to go?

I just saw your comments from the weekend, weird that you didn't see the changes... I'm sure I changed those things.

**Pull request closed without merging**